File: q1838/q1838.java
class Solution {
    public int maxFrequency(int[] nums, int k) {
        Arrays.sort(nums);

        int left = 0;
        long windowSum = 0;
        int res = 0;
        for (int right = 0; right < nums.length; right++) {

            // expand the window
            long newTarget = nums[right];
            windowSum += newTarget;

            // shrink the window to ensure tha conditions are not broken
            while (windowSum + k < (right - left + 1) * newTarget) {
                windowSum -= nums[left];
                left++;
            }

            // update the result after the
            res = Math.max(res, right - left + 1);
        }

        return res;
    }
}

File: q1838/q1838.cpp
class Solution
{
public:
    int maxFrequency(vector<int> &nums, int k)
    {
        sort(nums.begin(), nums.end());
        int left = 0;
        long windowSum = 0;
        int res = 0;

        for (int right = 0; right < nums.size(); right++)
        {
            // expand the window
            long num = nums[right];
            windowSum += num;

            // shrink the window until condition met
            while (windowSum + k < (right - left + 1) * num)
            {
                windowSum -= nums[left];
                left++;
            }

            // update the res
            res = max(right - left + 1, res);
        }
        return res;
    }
};